src/GNDStk/gnds-keyword-read.hpp

```diff
@@ -0,0 +1,132 @@
+
+// Note that our default gnds::read() just uses stream input. Generally, then,
+// a user would just make a run-of-the-mill operator>>, as they'd probably
+// wish and expect to. However, we chose not have gnds::read() layered around
+// normal stream input, just in case somebody wants to avoid foisting stream
+// input capabilities on a type that perhaps shouldn't have it. For example,
+// we may not really want stream input for C++ containers.
+
+
+
+// -----------------------------------------------------------------------------
+// read(istream,T)
+// default: use operator>>
+// -----------------------------------------------------------------------------
+
+// default
+template<class T>
+inline void read(std::istream &is, T &value)
+{
+   is >> value;
+}
+
+
+// ------------------------
+// sequence containers,
+// except array and
+// forward_list
+// ------------------------
+
+#define gnds_read(container) \
+   template<class T, class Alloc> \
+   inline void read(std::istream &is, std::container<T,Alloc> &value) \
+   { \
+      value.clear(); \
+      T v; \
+      while (is >> v) \
+         value.push_back(v); \
+   }
+
+gnds_read(deque)
+gnds_read(list)
+gnds_read(vector)
+
+#undef gnds_read
+
+
+
+// -----------------------------------------------------------------------------
+// read(string,T)
+// default: make istringstream from string, then use read(istream,T) above
+// -----------------------------------------------------------------------------
+
+// default
+template<class T>
+inline void read(const std::string &str, T &value)
+{
+   std::istringstream iss(str);
+   read(iss,value);
+}
+
+// string
+inline void read(const std::string &str, std::string &value)
+{
+   value = str;
+}
+
+// bool
+inline void read(const std::string &str, bool &value)
+{
+   if (
+      str == "1" || str == "t"     || str == "true"  ||
+      str == "T" || str == "True"  || str == "TRUE"  )
+      value = true;
+   else if (
+      str == "0" || str == "f"     || str == "false" ||
+      str == "F" || str == "False" || str == "FALSE" )
+      value = false;
+   else {
+      // fixme do something better than this
+      assert(false);
+      value = false;
+   }
+}
+
+// miscellaneous
+// string-to-T specializations that may be faster than our default
+#define gnds_read(fun,T) \
+   inline void read(const std::string &str, T &value) \
+   { \
+      value = std::fun(str); \
+   }
+
+gnds_read(stod, double)
+gnds_read(stof, float)
+gnds_read(stoi, int)
+gnds_read(stol, long)
+gnds_read(stold, long double)
+gnds_read(stoll, long long)
+gnds_read(stoul, unsigned long)
+gnds_read(stoull, unsigned long long)
+
+#undef gnds_read
+
+
+
+// -----------------------------------------------------------------------------
+// read(node,T)
+// default: error
+// -----------------------------------------------------------------------------
+
+// default
+// Doesn't know what to do; should be overridden
+// fixme Consider having *no* default for read(node,...)
```

Comment:
This could be a discussion about what to do in this particular location, or a much more general conversation. My general experience is that some people love the C++ exception handling system, some people avoid it like the plague. Exception handling almost always introduces a performance hit, both space-wise and time-wise. This even includes cases where someone doesn't use exception handling explicitly anywhere in their code. If they call some std:: (or any) function that uses it in any form, the compiler has to put extra junk all over the place in the entire call graph of the code that calls (or can call) the offending function. This is why compilers provide options like --no-exceptions to completely turn off exception handling at a compile-time level. Over the years, I've seen compilers get better about optimizing stuff that may have exception-handling stuff lurking here and there, but we're still talking about some level of performance hit, and it may or may not be important in a code like this. Aside from that question, some users like other stuff than exceptions to happen. Maybe they're quickly developing a fast-and-dirty app and just want the thing to bomb out (e.g. assertion failure, or error message and exit()) if anything goes wrong. Some people like a true/false return, or something of that nature, if something goes wrong. I see that your convention in the overarching project's set of libraries is to use C++ exceptions, and I'll certainly honor that convention as work proceeds. (In many places I have stuff like assertion failures at the moment; those were meant all along as placeholders, so I could write material quickly, and get the code out there.) If we're open to considering some options that might make some users really happy, I think I can easily write some code that would package up errors in some consolidated place, which then defaults to a "throw exception" behavior, per project policy, but also support some sort of global flag that would allow a user who despises exceptions to switch to another methodology. This is one of the ideas behind that diagnostic-message and ANSI-color-escape-sequence business I put in, and that you mentioned in other comments. Another offline discussion to have sometime.

Reply:
The code in question was largely reworked a while back, and has what I think is the right thing to do at this time. Regarding diagnostics, lots of "fixme"s throughout the code need to be visited and turned into real diagnostics. I've just been focusing on other things so far.